### PR TITLE
integration test: add waits for RBACs

### PR DESCRIPTION
**Pull request by listx**
Sometimes, the rbac rules don't get created before Sinker is deployed (this can happen in CI). So we break up the definitions like the other services.

Instead of creating new files, we symlink back to the configurations used in production, like we do for other services we test.

### CODE REVIEW

**Comment by matthyx** [comment]
/lgtm
/hold
if you need it quickly you can unhold, otherwise let's wait for Chao

**Review by chaodaiG — APPROVED** [review]

**Comment by listx** [comment]
OK so the test was still flaky because the application of these configs was done in an async manner with backgrounded jobs (`&` symbol in bash). I've removed job backgrounding in a new commit so this should be good.

But as another added measure, I will run `/retest` 10 times in a row just to get some added confidence that this PR actually does reduce flakiness. It's going to be a little noisy so I'll mark this as a WIP to hopefully reduce email noise during this period of manual testing.

/hold

**Comment by listx** [comment]
> @listx: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-test-infra-integration	[73aa534](https://github.com/kubernetes/test-infra/commit/73aa53493f99978c9267ccada650b0a34b9c4b34)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602782151462359040)	true	`/test pull-test-infra-integration`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=test-infra&pr=28236). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Alistx). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/test-infra/issues?q=is:issue+is:open) when you hit one in your PR.

This failed because apparently there's a small bug with our configuration for sinker not having a `protocol: ...` field for the service. I've updated this config in the latest commit.

**Comment by listx** [comment]
1/10: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602786919312265216

/test pull-test-infra-integration

**Comment by listx** [comment]
2/10: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602792437619625984

1/10: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602786919312265216

/test pull-test-infra-integration

**Comment by listx** [comment]
3/10: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602826145856753664

/test pull-test-infra-integration

**Comment by listx** [comment]
4/10: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602839643970080768

/test pull-test-infra-integration

**Comment by listx** [comment]
6/10: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602850007541616640


5/10: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1602846490324111360

/test pull-test-infra-integration

**Comment by listx** [comment]
I've realized that I should include some more explicit waits into the startup script as well. See https://stackoverflow.com/questions/57115602/how-to-kubectl-wait-for-crd-creation.

Basically we should be inserting a lot more waits for every single resource we create and care about to guarantee that the startup process completes correctly, in the correct sequence. This is because `kubectl apply` itself is asynchronous (regardless of whether we background the job or not). The fact that the test runs above have been passing is likely due to the fact that we've added just enough delays (by virtue of looping over more items) before we deploy Sinker.

If we add these explicit waits, I'll be comfortable getting this merged without running this 10x times to collect data.

**Comment by listx** [comment]
Interesting... from https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/28036/pull-test-infra-integration/1603196148191858688 I'm seeing:

```
...
rolebinding.rbac.authorization.k8s.io/prow-controller-manager serverside-applied
rolebinding.rbac.authorization.k8s.io/prow-controller-manager serverside-applied
service/prow-controller-manager serverside-applied
deployment.apps/prow-controller-manager serverside-applied
role.rbac.authorization.k8s.io/sinker serverside-applied
Error from server: etcdserver: request timed out
Error from server (InternalError): Internal error occurred: resource quota evaluation timed out
```

Some internet searches seem to suggest that this is a resource contention issue come from Kubernetes. For the `etcdserver: request timed out` one I saw suggestions to upgrade to an SSD. If I go into the [logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/test-infra/28036/pull-test-infra-integration/1603196148191858688/artifacts/kind_logs/kind-prow-integration-control-plane/pods/kube-system_etcd-kind-prow-integration-control-plane_874491576576018d181fdd0c07214bd7/etcd/0.log) for the etcd container, I see:

```
...
2022-12-15T01:22:21.155562141Z stderr F 2022-12-15 01:22:21.155263 W | etcdserver/api/etcdhttp: /health error; QGET failed etcdserver: request timed out (status code 503)
2022-12-15T01:22:22.186277096Z stderr F 2022-12-15 01:22:22.185809 W | etcdserver: timed out waiting for read index response (local node might have slow network)
2022-12-15T01:22:22.186337534Z stderr F 2022-12-15 01:22:22.186042 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/kube-apiserver-kind-prow-integration-control-plane\" " with result "error:etcdserver: request timed out" took too long (12.702731172s) to execute
2022-12-15T01:22:22.18634875Z stderr F 2022-12-15 01:22:22.186078 W | etcdserver: read-only range request "key:\"/registry/clusterrolebindings/\" range_end:\"/registry/clusterrolebindings0\" count_only:true " with result "error:etcdserver: request timed out" took too long (13.597031074s) to execute
2022-12-15T01:22:22.186383759Z stderr F 2022-12-15 01:22:22.186134 W | etcdserver: read-only range request "key:\"/registry/clusterroles/\" range_end:\"/registry/clusterroles0\" count_only:true " with result "error:etcdserver: request timed out" took too long (13.687671168s) to execute
2022-12-15T01:22:22.18639002Z stderr F 2022-12-15 01:22:22.186197 W | etcdserver: read-only range request "key:\"/registry/flowschemas/exempt\" " with result "error:etcdserver: request timed out" took too long (9.299432796s) to execute
2022-12-15T01:22:22.18639651Z stderr F 2022-12-15 01:22:22.186222 W | etcdserver: read-only range request "key:\"/registry/ingressclasses/\" range_end:\"/registry/ingressclasses0\" count_only:true " with result "error:etcdserver: request timed out" took too long (8.018279301s) to execute
2022-12-15T01:22:22.186875196Z stderr F 2022-12-15 01:22:22.186565 W | etcdserver: read-only range request "key:\"/registry/namespaces/kube-system\" " with result "error:etcdserver: request timed out" took too long (8.942955763s) to execute
2022-12-15T01:22:22.56490616Z stderr F 2022-12-15 01:22:22.564582 W | wal: sync duration of 7.379122566s, expected less than 1s
...
```

Maybe we need a beefier cluster for the integration test job?

In lieu of that, I can try adding in even more pauses to make the various resource creation requests more staggered in the hopes of not overwhelming the API server (or etcd). Looking at the first paste, I can tell that this error happened right after we requested the creation of a large number of resources. So I think I can add in some pauses in between each of those to  purposely slow things down.

**Comment by listx** [comment]
The latest push appears to have improved the health of etcd. Previously I posted `wal: sync duration of 7.379122566s, expected less than 1s` but if I grep for it in the [current test run](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/test-infra/28236/pull-test-infra-integration/1603257375647600640/artifacts/kind_logs/kind-prow-integration-control-plane/pods/kube-system_etcd-kind-prow-integration-control-plane_874491576576018d181fdd0c07214bd7/etcd/0.log) (another [example](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/test-infra/28036/pull-test-infra-integration/1603258794589032448/artifacts/kind_logs/kind-prow-integration-control-plane/pods/kube-system_etcd-kind-prow-integration-control-plane_874491576576018d181fdd0c07214bd7/etcd/0.log)) I see numbers around 1.1 to 1.5 seconds, which is definitely less stressful for etcd.

I think the key is that we wait for the creation of the massive Prow Job CRD (21K lines, or 1.5MB of text data) to finish before asking the K8s API server to create more things (whereas before we did things concurrently). Perhaps the best part is that the changes in this PR which serialize the setup of resources does not appear to have had any negative performance impact on the timing of the tests. For example, the two links I posted in this comment took 12m4s and 12m44s, which are comparable to previous timings.

File: prow/test/integration/lib.sh [hunk 1]
```diff
@@ -210,6 +275,71 @@ function wait_for_readiness() {
...
+    echo >&2 "wanted a CSV of exactly 2 values, with the syntax '<NAME>,<NAMESPACE>'; got '${arg}'"
+  fi
+
+  name="${args[0]}"
+  namespace="${args[1]}"
+
+  echo >&2 "waiting for CRD ${name} in namespace '${namespace}'..."
+
+  for _ in $(seq 1 10); do
+  #WAIT_FOR_CRD_ kubectl -n default wait --for condition=established --timeout=60s crd prowjobs.prow.k8s.io
```
Comment (listx) [inline prow/test/integration/lib.sh:332]: I forgot to delete this line. Will delete

**Comment by listx** [comment]
@chaodaiG or @cjwagner can you LGTM?

**Review by chaodaiG — APPROVED** [review]

**Pull request merged**